### ISSUE TITLE
TaskGroup does not support dynamically generated tasks

### ISSUE THREAD

**Issue by turbaszek** [apache/airflow#12309]
**Apache Airflow version**:
2.0 / master

**Environment**:

breeze

**What happened**:

Using this DAG:

```py
from airflow.operators.bash import BashOperator
from airflow.operators.python import task
from airflow.models import DAG
from airflow.utils.task_group import TaskGroup

@task
def show():
    print("Cats are awesome!")

with DAG(
    "using_task_group",
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
) as dag3:
    start_task = BashOperator(
        task_id="start_task",
        bash_command="echo start",
    )

    end_task = BashOperator(
        task_id="end_task",
        bash_command="echo end",
    )

    with TaskGroup(group_id="show_tasks") as tg1:
        previous_show = show()
        for _ in range(100):
            next_show = show()
            previous_show >> next_show
            previous_show = next_show
```
I get:
```
Broken DAG: [/files/dags/test.py] Traceback (most recent call last):
  File "/opt/airflow/airflow/models/baseoperator.py", line 410, in __init__
    task_group.add(self)
  File "/opt/airflow/airflow/utils/task_group.py", line 140, in add
    raise DuplicateTaskIdFound(f"Task id '{key}' has already been added to the DAG")
airflow.exceptions.DuplicateTaskIdFound: Task id 'show_tasks.show' has already been added to the DAG
```

If I remove the task group the task are generated as expected.

**What you expected to happen**:

I expect to be able to generate tasks dynamically using TaskGroup and task decoratos.

**How to reproduce it**:

Use the DAG from above.

**Anything else we need to know**:

N/A

**Comment by turbaszek** [comment]
CC @yuqian90 @casassg

**Comment by turbaszek** [comment]
Minimal example to replicate it:
```py
from airflow.models import DAG
from airflow.operators.python import task
from airflow.utils.dates import days_ago
from airflow.utils.state import State
from airflow.utils.task_group import TaskGroup

@task
def show():
    print("cats are awesome")

with DAG(
    "using_task_group",
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
) as dag:

    with TaskGroup(group_id="tasks_1") as tg1:
        show()
        show()


if __name__ == '__main__':
    dag.clear(dag_run_state=State.NONE)
    dag.run()
```

**Comment by potiuk** [comment]
Not sure how complex it is and whether we can manage it for beta3/4/5 @yuqian90 @casassg ?

**Comment by casassg** [comment]
Looking at @turbaszek code, that should address this issue

**Comment by rajib2k5** [comment]
Is TaskGroup supported in Airflow 1.10.10?

**Comment by potiuk** [comment]
> Is TaskGroup supported in Airflow 1.10.10?

No

**Comment by ssiddiqui-apixio** [comment]
I am using airflow version 2.0.1 and i can still reproduce this error. When I look at the code in my python3.7/site-packages/airflow/ folders, i see code that is consistent with the changes made here: https://github.com/apache/airflow/pull/12312 .

**Comment by ashb** [comment]
@ssiddiqui-apixio Can you give us a minimal reproduction case please?

**Comment by ssiddiqui-apixio** [comment]
sure, @ashb 

This is the case which works.
`
from airflow.models import DAG
from airflow.operators.python import task
from airflow.utils.dates import days_ago
from airflow.utils.task_group import TaskGroup

@task
def show():
    print("cats are awesome")

with DAG(
    "using_task_group",
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
) as dag3:

    with TaskGroup(group_id="tasks_1") as tg1:
        for _ in range(10):
            BashOperator(task_id=f'processing_{_}', bash_command='ls')
`

This case fails. The difference here is that in the taskGroup, I am calling a task via a method decorator. 

`from airflow.models import DAG
from airflow.operators.python import task
from airflow.utils.dates import days_ago
from airflow.utils.task_group import TaskGroup

@task
def show():
    print("cats are awesome")

with DAG(
    "using_task_group",
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
) as dag3:

    with TaskGroup(group_id="tasks_1") as tg1:
        for _ in range(10):
            show()`

This scenario results in the following error:

> 
(airflow) ssiddiqui@siddiquiMac ~ % airflow dags test using_task_group 2021-03-19T19:20:03
[2021-03-23 14:29:32,716] {dagbag.py:448} INFO - Filling up the DagBag from /Users/ssiddiqui/dev/airflow/spark/data_v2/stg
/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/future/standard_library/__init__.py:65 DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
[2021-03-23 14:29:33,214] {dagbag.py:305} ERROR - Failed to import: /Users/ssiddiqui/dev/airflow/spark/data_v2/stg/bug_example.py
Traceback (most recent call last):
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/models/dagbag.py", line 302, in _load_modules_from_file
    loader.exec_module(new_module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/ssiddiqui/dev/airflow/spark/data_v2/stg/bug_example.py", line 79, in <module>
    show()
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/operators/python.py", line 294, in factory
    **kwargs,
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/models/baseoperator.py", line 89, in __call__
    obj: BaseOperator = type.__call__(cls, *args, **kwargs)
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/utils/decorators.py", line 94, in wrapper
    result = func(*args, **kwargs)
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/operators/python.py", line 170, in __init__
    super().__init__(**kwargs)
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/utils/decorators.py", line 94, in wrapper
    result = func(*args, **kwargs)
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/models/baseoperator.py", line 410, in __init__
    task_group.add(self)
  File "/Users/ssiddiqui/.pyenv/versions/3.7.3/envs/airflow/lib/python3.7/site-packages/airflow/utils/task_group.py", line 140, in add
    raise DuplicateTaskIdFound(f"Task id '{key}' has already been added to the DAG")
airflow.exceptions.DuplicateTaskIdFound: Task id 'tasks_1.show__1' has already been added to the DAG
>

**Comment by ahayasic** [comment]
@ssiddiqui-apixio I'm having the same problem on Airflow 2.0.2. Any chance of the Issue being reopened for a new solution?

**Comment by sjacquot** [comment]
I'm still having the problem on Airflow 2.1.2

_Broken DAG: [/usr/local/airflow/dags/dags/auto/bronze/xywyxw_full_bronze_dag.py] Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/baseoperator.py", line 524, in __init__
    task_group.add(self)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/task_group.py", line 158, in add
    raise DuplicateTaskIdFound(f"Task id '{key}' has already been added to the DAG")
airflow.exceptions.DuplicateTaskIdFound: Task id 'xywyxw_wwww_full_bronze-20220217-104624' has already been added to the DAG_

Can you please reopen this issue to track the problem to a new solution ?

**Comment by potiuk** [comment]
I believe it's just wrong way of using decorated tasks. 

But the issue is that you are adding the same issue again with the same taks_id. You shoudl change task_id when you are adding tasks mutiple times. This problem is not fixable IMHO because of the way how decorators work - the function for the resulting task will always have the same id and (as far as I know how decorarots work) this cannot be changed. 

If you want to have a loop of tasks added use PythonOperator and use your function as callable. 

Unless someone objects and tells that this can be solved differently i consider this one closed.

**Comment by ashb** [comment]
(That `test_multiple_calls_in_task_group` was added in #14709 was released in 2.1.0.)

**Comment by potiuk** [comment]
Ah cool. TIL.
